common/CommonAPI.h

```diff
@@ -722,23 +711,281 @@ class CommonAPI
+	{
+		uint64_t width : 14;
+		uint64_t height : 14;
+		uint64_t resourceIx : 8; // (Frame in flight)
+		uint64_t frameIx : 28; // (Total amount of frames rendered so far / frame index)
+	};
+
+	std::atomic_uint64_t m_lastPresentedFrame;
+
+	PresentedFrameInfo getLastPresentedFrame()
+	{
+		uint64_t value = m_lastPresentedFrame.load();
+		PresentedFrameInfo frame;
+		frame.width = value >> 50;
+		frame.height = (value >> 36) & ((1 << 14) - 1);
+		frame.resourceIx = (value >> 28) & ((1 << 8) - 1);
+		frame.frameIx = value & ((1 << 28) - 1);
+
+		return frame;
+	}
+
+	void setLastPresentedFrame(PresentedFrameInfo frame)
+	{
+		uint64_t value = 0;
+		value |= frame.width << 50;
+		value |= frame.height << 36;
+		value |= frame.resourceIx << 28;
+		value |= frame.frameIx;
+		m_lastPresentedFrame.store(value);
+	}
+
+	virtual void onCreateResourcesWithTripleBufferTarget(nbl::core::smart_refctd_ptr<nbl::video::IGPUImage>& image, uint32_t bufferIx) {}
+
+	nbl::video::IGPUImage* getTripleBufferTarget(
+		uint32_t frameIx, uint32_t w, uint32_t h, 
+		nbl::asset::E_FORMAT surfaceFormat, 
+		nbl::core::bitflag<nbl::asset::IImage::E_USAGE_FLAGS> imageUsageFlags)
+	{
+		uint32_t bufferIx = frameIx % 2;
+		auto& image = m_tripleBufferRenderTargets.begin()[bufferIx];
+
+		if (!image || image->getCreationParameters().extent.width < w || image->getCreationParameters().extent.height < h)
+		{
+			auto logicalDevice = getLogicalDevice();
+			nbl::video::IGPUImage::SCreationParams creationParams;
+			creationParams.type = nbl::asset::IImage::ET_2D;
+			creationParams.samples = nbl::asset::IImage::ESCF_1_BIT;
+			creationParams.format = surfaceFormat;
+			creationParams.extent = { w, h, 1 };
+			creationParams.mipLevels = 1;
+			creationParams.arrayLayers = 1;
+			creationParams.usage = imageUsageFlags | nbl::asset::IImage::EUF_TRANSFER_SRC_BIT;
+
+			image = logicalDevice->createImage(std::move(creationParams));
+			auto memReqs = image->getMemoryReqs();
+			memReqs.memoryTypeBits &= logicalDevice->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+			logicalDevice->allocate(memReqs, image.get());
+
+			onCreateResourcesWithTripleBufferTarget(image, bufferIx);
+		}
+
+		return image.get();
+	}
+#endif
 public:
 	GraphicalApplication(
 		const std::filesystem::path& _localInputCWD,
 		const std::filesystem::path& _localOutputCWD,
 		const std::filesystem::path& _sharedInputCWD,
 		const std::filesystem::path& _sharedOutputCWD
-	) : nbl::system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
-	void recreateSurface() override
+	) : nbl::system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD),
+		CommonAPI::CommonAPIEventCallback(nullptr, nullptr),
+		m_qRetiredSwapchainResources(),
+		m_imageSwapchainIterations{}
+	{}
+
+	std::unique_lock<std::mutex> recreateSwapchain(
+		uint32_t w, uint32_t h, 
+		nbl::video::ISwapchain::SCreationParams& swapchainCreationParams, 
```

Comment:
shouldn't the creation parameters be persistent (immutable) and therefore not an argument to this method?

Reply:
its not like you're gonna change formats, min image count etc.

Perhaps we should move `CommonAPI::createSwapchain` method to `IGraphcalApplication`.

Then only ask for things _that can feasibly change_ in `recreateSwapchain` which are:
- width
- height
- preTransform?

Add a `ISwapchain::getCreationParams()`, patch them up with changes, and use that.

Reply:
maybe make this a method of `CResizableSwapchainUtils`